src/Tools/ExternalAccess/OmniSharp.CSharp/Completion/OmniSharpCompletionProviderNames.cs

```diff
@@ -2,17 +2,15 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using Microsoft.CodeAnalysis.CSharp.Completion.Providers;
-
 namespace Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.CSharp.Completion
 {
     internal static class OmniSharpCompletionProviderNames
     {
-        internal static string ObjectCreationCompletionProvider = typeof(ObjectCreationCompletionProvider).FullName;
-        internal static string OverrideCompletionProvider = typeof(OverrideCompletionProvider).FullName;
-        internal static string PartialMethodCompletionProvider = typeof(PartialMethodCompletionProvider).FullName;
-        internal static string InternalsVisibleToCompletionProvider = typeof(InternalsVisibleToCompletionProvider).FullName;
-        internal static string TypeImportCompletionProvider = typeof(TypeImportCompletionProvider).FullName;
-        internal static string ExtensionMethodImportCompletionProvider = typeof(ExtensionMethodImportCompletionProvider).FullName;
+        internal const string ObjectCreationCompletionProvider = "Microsoft.CodeAnalysis.CSharp.Completion.Providers.ObjectCreationCompletionProvider";
+        internal const string OverrideCompletionProvider = "Microsoft.CodeAnalysis.CSharp.Completion.Providers.OverrideCompletionProvider";
+        internal const string PartialMethodCompletionProvider = "Microsoft.CodeAnalysis.CSharp.Completion.Providers.PartialMethodCompletionProvider";
+        internal const string InternalsVisibleToCompletionProvider = "Microsoft.CodeAnalysis.CSharp.Completion.Providers.InternalsVisibleToCompletionProvider";
+        internal const string TypeImportCompletionProvider = "Microsoft.CodeAnalysis.CSharp.Completion.Providers.TypeImportCompletionProvider";
+        internal const string ExtensionMethodImportCompletionProvider = "Microsoft.CodeAnalysis.CSharp.Completion.Providers.ExtensionMethodImportCompletionProvider";
```

Comment:
These should all be `static readonly` to ensure they are not embedded as literals into external code.

Reply:
I'm going to keep these as constants. I don't think we have a real problem here: OmniSharp controls what version of Roslyn they take, no one is going to be upgrading it to a new version. We have tests now to ensure that these reflect the real names of the types, and the code in O# that uses these names has made pretty significant usage of pattern matching.

Reply:
The external access layer ships with Roslyn itself. The value of these strings is allowed to change as long as the following two requirements hold:

1. The name of the field stays the same
2. The value of the field is the new type name for the completion provider with the same concept

Embedding the strings in a binary outside of Roslyn breaks the ability to change the value in account for (2).

Reply:
In other words, Roslyn is allowed to change the constant value here _and_ create an assembly binding redirection that forces the new value to appear at runtime. Any code referencing these as `const` would no longer behave correctly.

Reply:
> In other words, Roslyn is allowed to change the constant value here and create an assembly binding redirection that forces the new value to appear at runtime.

There is no one to create such a binding redirect. OmniSharp is the host process here, and controls the final version of Roslyn being used. It's never going to be using a different version of the external access library than the rest of roslyn. While I get the theoretical concern here, I don't believe it matters for this external access library and I'm going to keep these as constants. Again, there is significant code in OmniSharp that _already_ uses these as constants, and I'm not going to rewrite this code to a wildly different form to conform to theoretical concerns. I'll just not use these fields and leave them as the OmniSharp-copied constants.

Reply:
> I'll just not use these fields and leave them as the OmniSharp-copied constants.

This would be an acceptable alternative for this case. If OmniSharp wants a unit test verifying that the constants are correct, the unit tests could either use reflection, or you could keep the `static readonly` values here in the external access layer and assert that they match the copied constants.

Reply:
> We sure don't expect or support TypeScript or F# external access where they are compiling against one but we're binding redirecting to others as well.

We not only support this, but it's the primary use case in deployment. The external access layer for TypeScript ships with _Roslyn_ components, not TypeScript, and Roslyn binding redirects whatever TypeScript compiled against to the current version it installed.

Reply:
> The external access layer for TypeScript ships with Roslyn components, not TypeScript, and Roslyn binding redirects whatever TypeScript compiled against to the current version it installed.

Sure, and I get that this is a concern with other components. It's explicitly _not_ a concern for this dll. OmniSharp is the host process, and controls the version of Roslyn being used. It's not going to be upgraded by external dependencies, and OmniSharp was designed to support scenario as a goal.

Reply:
can i ask a clarifying question: what value do we get from constants here?  they seem strictly more brittle than static readonlies.

Reply:
> can i ask a clarifying question: what value do we get from constants here? they seem strictly more brittle than static readonlies.

Constants can appear in the arms of a switch expression, but `static readonly` cannot.

Reply:
Or any pattern matches: as mentioned, the completion code in OmniSharp makes quite a bit of use of pattern matching over these names today.

Reply:
Alright, I'll revert this. But ultimately, I'm not going to be changing the code in omnisharp here. It just means we're going to have constants there and not use these fields.

Reply:
@333fred maybe a situation to keep on the back of your mind it case future pattern matching improvements come through LDM 😄